dask/dataframe/tests/test_dataframe.py

```diff
@@ -5109,3 +5109,38 @@ def reduce_fn(series):
         .compute()[0]
     )
     assert result == {"x": 14, "y": 64}
+
+
+@pytest.fixture
+def demo_ts():
+    return dd.demo.make_timeseries()
+
+
+def test_set_index_with_dask_dt_index_gh8359(demo_ts):
+    ## was failing, now works
+    # https://github.com/dask/dask/issues/8359
+    new_index = demo_ts.index.dt.floor("D")
+    demo_ts.set_index(new_index)
+
+
+def test_set_index_with_dask_range_index_gh8359(demo_ts):
```

Comment:
Rather than an individual test for several dtypes, we may be able to parameterize across different indices and write a single test.